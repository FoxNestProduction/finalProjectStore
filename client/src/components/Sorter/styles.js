export const stylesWrap = {
  display: 'flex',
  justyfyContent: 'column',
  fontFamily: 'fontPoppins',
  width: {
    mobile: '316px',
    tablet: '268px',
    desktop: '377px',
  },
  borderRadius: '17px',
  bgcolor: '#ffffff',
  color: 'text.secondary',
  p: {
    mobile: '27px',
    tablet: '23px',
    desktop: '32px',
  },
};

export const stylesTitle = {
  fontSize: {
    mobile: '18px',
    tablet: '15px',
    desktop: '20px',
  },
  fontFamily: 'fontPoppins',
  fontWeight: '600',
  color: '#000000',
};

export const stylesSortBtn = {
  color: 'text.secondaryGray',
  fontWeight: '500',
  fontFamily: 'fontPoppins',
  whiteSpace: 'nowrap',
  fontSize: {
    mobile: '13px',
    tablet: '11px',
    desktop: '16px',
  },
  '&.MuiButtonBase-root': {
    border: '0',
  },
  px: '0',
  '&.Mui-selected': {
    color: 'primary.main',
    bgcolor: '#ffffff',
  },
};

export const stylesCategoryIconsWrap = {
  display: 'flex',
  justyfyContent: 'column',
  gap: {
    mobile: '10px',
    tablet: '9px',
    desktop: '13px',
  },
  borderRadius: '15px',
  bgcolor: '#ffffff',
  mt: '20px',
};

export const stylesToggleButton = {
  p: '0',
  width: {
    mobile: '55px',
    tablet: '47px',
    desktop: '66px',
  },
  height: {
    mobile: '80px',
    tablet: '68px',
    desktop: '96px',
  },
  borderRadius: '15px',
  bgcolor: '#FFDE8A',
  '&:hover': {
    bgcolor: '#ffffff',
    border: '4px solid rgba(251, 71, 29, 0.50)',
  },
  '&.Mui-selected': {
    color: '#FB471D',
    bgcolor: '#F7C5BA',
    border: '4px solid rgba(251, 71, 29, 0.50)',
    '&:hover': {
      bgcolor: '#ffffff',
      // border: '4px solid rgba(251, 71, 29, 0.50)',
    },
  },
};

export const stylesCategoryItem = {
  display: 'flex',
  justyfyContent: 'space-around',
  alignItems: 'center',
  gap: {
    mobile: '8px',
    tablet: '7px',
    desktop: '10px',
  },
  '& .MuiCardMedia-root': {
    width: {
      mobile: '29px',
      tablet: '25px',
      desktop: '35px',
    },
    height: {
      mobile: '29px',
      tablet: '25px',
      desktop: '35px',
    },
  },
  '& .MuiTypography-root': {
    fontFamily: 'fontPoppins',
    fontWeight: '500',
    fontSize: {
      mobile: '11px',
      tablet: '9px',
      desktop: '12px',
    },
  },
};

export const stylesSlider = {
  mt: {
    mobile: '15px',
    tablet: '12px',
    desktop: '17px',
  },
  fontSize: {
    mobile: '9px',
    tablet: '8px',
    desktop: '11px',
  },
  '.MuiSlider-markLabel': {
    fontFamily: 'fontPoppins',
  },
  fontWeight: '500',
  '.MuiSlider-valueLabel': {
    fontFamily: 'fontPoppins',
    fontSize: '7px',
    lineHeight: '12px',
    p: '0.25rem 0.5rem',
  },
};

export const stylesBtn = {
  color: 'text.primaryLight',
  bgcolor: 'primary.main',
  fontFamily: 'fontPoppins',
  '.MuiButton-root': {},
  '&:hover': {
    bgcolor: 'primary.hover',
    color: 'text.primaryLight',
    borderColor: 'primary.hover',
  },
  mt: {
    mobile: '32px',
    tablet: '26px',
    desktop: '40px',
  },
};
