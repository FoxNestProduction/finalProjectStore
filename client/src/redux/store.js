import { configureStore, getDefaultMiddleware, combineReducers } from '@reduxjs/toolkit';
// todo: правила для redux-persist
// eslint-disable-next-line import/no-extraneous-dependencies
import storage from 'redux-persist/lib/storage';
// eslint-disable-next-line import/no-extraneous-dependencies
import { persistReducer } from 'redux-persist';
import modalSlice from './slices/modalSlice';
import productsSlice from './slices/productsSlice';
import authorizationSlice from './slices/authorizationSlice';
import userSlice from './slices/userSlice';
import partnersSlice from './slices/partnersSlice';
import searchSlice from './slices/searchSlice';
import errorSlice from './slices/errorSlice';
import reviewsSlice from './slices/reviewsSlice';
import favouriteSlice from './slices/favouriteSlice';

const authPersistConfig = {
  key: 'authorization',
  version: 1,
  storage,
};

const userPersistConfig = {
  key: 'user',
  version: 1,
  storage,
};

const searchPersistConfig = {
  key: 'search',
  version: 1,
  storage,
};
const favouritePersistConfig = {
  key: 'favorites',
  version: 1,
  storage,
};

/*
    todo: - ЗВЕРНІТЬ УВАГУ!
      якщо вам потрібно щоб ваш стейт не зникав після перезавантаження
      то юзайте підхід як authorization і user,
      якщо ж ні то варіант написання як для modal всередині reducer
      Якщо що пишіть @Ihor_Kacher
*/
const reducer = combineReducers({
  error: errorSlice,
  modal: modalSlice,
  partners: partnersSlice,
  authorization: persistReducer(authPersistConfig, authorizationSlice),
  user: persistReducer(userPersistConfig, userSlice),
  products: productsSlice,
  search: persistReducer(searchPersistConfig, searchSlice),
  reviews: reviewsSlice,
  favourites: persistReducer(favouritePersistConfig, favouriteSlice),
});

const middleware = getDefaultMiddleware({
  serializableCheck: false,
});

const store = configureStore({
  reducer,
  middleware,
});

export default store;
