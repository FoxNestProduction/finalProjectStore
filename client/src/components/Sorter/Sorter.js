import React, { memo } from 'react';
import { Box, MenuItem, TextField } from '@mui/material';
import { useTranslation } from 'react-i18next';
import PropTypes from 'prop-types';
import { shallowEqual, useDispatch, useSelector } from 'react-redux';
import { stylesSortSelect } from '../ListItems/styles';
import { setFilterParams } from '../../redux/slices/filterSlice';

const Sorter = ({ type }) => {
  const dispatch = useDispatch();
  const { i18n, t } = useTranslation();

  const filterParams = useSelector((state) => state.filter.filterParams, shallowEqual);

  const setSelectedValue = (sort) => {
    switch (sort) {
      case '+currentPrice': return 'Price UP';
      case '-currentPrice': return 'Price DOWN';
      case '+rating': return 'Rating UP';
      case '-rating': return 'Rating DOWN';
      default: return 'Default';
    }
  };

  let currencies;

  if (type === 'partners') {
    currencies = [
      {
        value: 'Rating UP',
        label: {
          en: 'Rating UP',
          ua: 'Рейтингом, від найменшого',
          pl: 'Rating UP',
        },
      },
      {
        value: 'Rating DOWN',
        label: {
          en: 'Rating DOWN',
          ua: 'Рейтингом, від найвищого',
          pl: 'Rating DOWN',
        },
      },
      {
        value: 'Default',
        label: {
          en: 'Default',
          ua: 'Замовчуванням',
          pl: 'Default',
        },
      },
    ];
  } else {
    currencies = [
      {
        value: 'Price UP',
        label: {
          en: 'Price UP',
          ua: 'Ціною, від найнижчої',
          pl: 'Price UP',
        },
      },
      {
        value: 'Price DOWN',
        label: {
          en: 'Price DOWN',
          ua: 'Ціною, від найвищої',
          pl: 'Price DOWN',
        },
      },
      {
        value: 'Rating UP',
        label: {
          en: 'Rating UP',
          ua: 'Рейтингом, від найнижчого',
          pl: 'Rating UP',
        },
      },
      {
        value: 'Rating DOWN',
        label: {
          en: 'Rating DOWN',
          ua: 'Рейтингом, від найвищого',
          pl: 'Rating DOWN',
        },
      },
      {
        value: 'Default',
        label: {
          en: 'Default',
          ua: 'Замовчуванням',
          pl: 'Default',
        },
      },
    ];
  }
  const handleSelectChangeSortBy = (event) => {
    switch (event.target.value) {
      case 'Price UP':
        dispatch(setFilterParams({ sort: '+currentPrice' }));
        break;

      case 'Price DOWN':
        dispatch(setFilterParams({ sort: '-currentPrice' }));
        break;

      case 'Rating UP':
        dispatch(setFilterParams({ sort: '+rating' }));
        break;

      case 'Rating DOWN':
        dispatch(setFilterParams({ sort: '-rating' }));
        break;

      default:
        dispatch(setFilterParams({ sort: '' }));
    }

    dispatch(setFilterParams({
      startPage: 1,
    }));
  };

  return (
    <Box sx={{ width: '100%', height: '40px', mb: '40px', paddingRight: '30px', textAlign: 'end' }}>
      <TextField
        sx={stylesSortSelect}
        id="standard-select-currency"
        size="small"
        select
        label={t('sorter.label')}
        defaultValue=""
        variant="standard"
        value={setSelectedValue(filterParams.sort)}
        onChange={handleSelectChangeSortBy}
      >
        {currencies.map((option) => (
          <MenuItem key={option.value} value={option.value}>
            {option.label[i18n.language]}
          </MenuItem>
        ))}
      </TextField>
    </Box>
  );
};

Sorter.propTypes = {
  type: PropTypes.string,
};

Sorter.defaultProps = {
  type: '',
};

export default memo(Sorter);
