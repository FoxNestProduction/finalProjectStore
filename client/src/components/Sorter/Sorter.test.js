import React from 'react';
import { fireEvent, render, screen } from '@testing-library/react';
import { Provider } from 'react-redux';
import Sorter from './Sorter';
import store from '../../redux/store';

const Component = (props) => {
  return (
    <Provider store={store}>
      <Sorter />
    </Provider>
  );
};

describe('Snapshot test', () => {
  test('should Sorter render', () => {
    const { asFragment } = render(<Component />);

    expect(asFragment()).toMatchSnapshot();
  });

  test('Sorter handles select change and dispatches actions for partners', async () => {
    render(<Component type="partners" />);
    const sortSelect = await screen.findByLabelText('Sort by');

    // Відкриваємо випадаючий список
    fireEvent.mouseDown(sortSelect);

    // Вибираємо значення
    const optionRatingUp = screen.getByText('Rating UP');
    fireEvent.click(optionRatingUp);

    const optionRatingDown = screen.getByText('Rating DOWN');
    fireEvent.click(optionRatingDown);

    const optionDefault = screen.getByText('Default');
    fireEvent.click(optionDefault);
  });

  test('Sorter handles select change and dispatches actions for products', async () => {
    render(<Component type="products" />);
    const sortSelect = await screen.findByLabelText('Sort by');

    // Відкриваємо випадаючий список
    fireEvent.mouseDown(sortSelect);

    // Вибираємо значення
    const optionRatingUp = screen.getByText('Rating UP');
    fireEvent.click(optionRatingUp);

    const optionRatingDown = screen.getByText('Rating DOWN');
    fireEvent.click(optionRatingDown);

    const optionPriceUp = screen.getByText('Price UP');
    fireEvent.click(optionPriceUp);

    const optionPriceDown = screen.getByText('Price DOWN');
    fireEvent.click(optionPriceDown);

    const optionDefault = screen.getByText('Default');
    fireEvent.click(optionDefault);
  });
});
