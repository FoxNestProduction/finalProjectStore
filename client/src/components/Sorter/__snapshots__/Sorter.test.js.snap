// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Snapshot test should Sorter render 1`] = `
<DocumentFragment>
  <div
    class="MuiBox-root css-fp9msj"
  >
    <div
      class="MuiFormControl-root MuiTextField-root css-1qi1fkt-MuiFormControl-root-MuiTextField-root"
    >
      <label
        class="MuiFormLabel-root MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-shrink MuiInputLabel-sizeSmall MuiInputLabel-standard MuiFormLabel-colorPrimary MuiFormLabel-filled MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-shrink MuiInputLabel-sizeSmall MuiInputLabel-standard css-1c2i806-MuiFormLabel-root-MuiInputLabel-root"
        data-shrink="true"
        for="standard-select-currency"
        id="standard-select-currency-label"
      >
        Sort by
      </label>
      <div
        class="MuiInputBase-root MuiInput-root MuiInput-underline MuiInputBase-colorPrimary MuiInputBase-formControl MuiInputBase-sizeSmall css-v4u5dn-MuiInputBase-root-MuiInput-root"
      >
        <div
          aria-expanded="false"
          aria-haspopup="listbox"
          aria-labelledby="standard-select-currency-label standard-select-currency"
          class="MuiSelect-select MuiSelect-standard MuiInputBase-input MuiInput-input MuiInputBase-inputSizeSmall css-jd1zyo-MuiSelect-select-MuiInputBase-input-MuiInput-input"
          id="standard-select-currency"
          role="button"
          tabindex="0"
        >
          Default
        </div>
        <input
          aria-hidden="true"
          aria-invalid="false"
          class="MuiSelect-nativeInput css-yf8vq0-MuiSelect-nativeInput"
          tabindex="-1"
          value="Default"
        />
        <svg
          aria-hidden="true"
          class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium MuiSelect-icon MuiSelect-iconStandard css-pqjvzy-MuiSvgIcon-root-MuiSelect-icon"
          data-testid="ArrowDropDownIcon"
          focusable="false"
          viewBox="0 0 24 24"
        >
          <path
            d="M7 10l5 5 5-5z"
          />
        </svg>
      </div>
    </div>
  </div>
</DocumentFragment>
`;
